linesearch.go

```diff
@@ -35,96 +39,152 @@ func (ls *LinesearchMethod) Init(loc *Location, xNext []float64) (EvaluationType
+		// previous location was not good enough to converge. So start the next
+		// linesearch and store the next evaluation point in loc.X.
+		return ls.initNextLinesearch(loc.X)
+
+	default:
+		// Store the result of the previously requested evaluation into ls.loc.
+		if ls.lastOp&FuncEvaluation != 0 {
+			ls.loc.F = loc.F
+		}
+		if ls.lastOp&GradEvaluation != 0 {
+			copy(ls.loc.Gradient, loc.Gradient)
+		}
+		if ls.lastOp&HessEvaluation != 0 {
+			ls.loc.Hessian.CopySym(loc.Hessian)
+		}
+		// Update the indicator of valid fields of ls.loc.
+		ls.eval |= ls.lastOp
+
+		if ls.nextMajor {
+			ls.nextMajor = false
+
+			// Linesearcher previously indicated that it had finished, but we
+			// needed to evaluate invalid fields of ls.loc. Now we have them and
+			// can announce MajorIteration.
 
-	if ls.iterType == MajorIteration {
-		// The linesearch previously signaled MajorIteration. Since we're here,
-		// it means that the previous location is not good enough to converge,
-		// so start the next linesearch.
-		return ls.initNextLinesearch(loc, xNext)
+			copyLocation(loc, &ls.loc)
+			ls.lastOp = MajorIteration
+			return ls.lastOp, nil
+		}
 	}
 
-	projGrad := floats.Dot(loc.Gradient, ls.dir)
-	if ls.Linesearcher.Finished(loc.F, projGrad) {
-		copy(xNext, loc.X)
-		// Check if the last evaluation evaluated all fields of Location.
-		ls.evalType = complementEval(loc, ls.evalType)
-		if ls.evalType == NoEvaluation {
-			// Location is complete and MajorIteration can be announced directly.
-			ls.iterType = MajorIteration
+	projGrad := floats.Dot(ls.loc.Gradient, ls.dir)
+	if ls.Linesearcher.Finished(ls.loc.F, projGrad) {
+		// Form an operation that evaluates invalid fields of ls.loc.
+		ls.lastOp = complementEval(&ls.loc, ls.eval)
+		if ls.lastOp == NoOperation {
+			// ls.loc is complete and MajorIteration can be announced directly.
+			copyLocation(loc, &ls.loc)
+			ls.lastOp = MajorIteration
 		} else {
-			// Location is not complete, evaluate its invalid fields in SubIteration.
-			ls.iterType = SubIteration
+			ls.nextMajor = true
 		}
-		return ls.evalType, ls.iterType, nil
+		return ls.lastOp, nil
 	}
 
-	// Line search not done, just iterate.
-	stepSize, evalType, err := ls.Linesearcher.Iterate(loc.F, projGrad)
+	step, op, err := ls.Linesearcher.Iterate(ls.loc.F, projGrad)
 	if err != nil {
-		ls.evalType = NoEvaluation
-		ls.iterType = NoIteration
-		return ls.evalType, ls.iterType, err
+		return ls.error(err)
+	}
+	if !op.isEvaluation() {
+		panic("linesearch: Linesearcher returned invalid operation")
 	}
 
-	floats.AddScaledTo(xNext, ls.x, stepSize, ls.dir)
-	// Compare the starting point for the current iteration with the next
-	// evaluation point to make sure that rounding errors do not prevent progress.
-	if floats.Equal(ls.x, xNext) {
-		ls.evalType = NoEvaluation
-		ls.iterType = NoIteration
-		return ls.evalType, ls.iterType, ErrNoProgress
+	if step != ls.lastStep {
```

Comment:
Would you mind swapping the if to have the equals case first? My first thought is "what happens if they're equal" and want to read below. Plus, that case is simpler and easy to digest.